homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +152,252 @@
 
-    def load_platforms(ipaddr):
-        platform_config = hass.data[DATA_YEELIGHT][ipaddr].config.copy()
-        platform_config[CONF_HOST] = ipaddr
-        platform_config[CONF_CUSTOM_EFFECTS] = config.get(DOMAIN, {}).get(
-            CONF_CUSTOM_EFFECTS, {}
-        )
-        load_platform(hass, LIGHT_DOMAIN, DOMAIN, platform_config, config)
-        load_platform(hass, BINARY_SENSOR_DOMAIN, DOMAIN, platform_config, config)
+        async def _initialize() -> None:
+            scanner = YeelightScanner(hass, entry)
+            unset = entry.add_update_listener(_async_update_listener)
+            hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = {
+                DATA_SCANNER: scanner,
+                DATA_UNSUB_UPDATE_LISTENER: unset,
+            }
 
-    dispatcher_connect(hass, DEVICE_INITIALIZED, load_platforms)
+            for component in PLATFORMS:
+                # Wait for platform setup is important. Make sure callbacks are set up before scanning.
+                await hass.config_entries.async_forward_entry_setup(entry, component)
+
+            await scanner.start_scan()
+
+        hass.async_create_task(_initialize())
+
+    else:
+
+        # Move options from data for imported entries
+        if not entry.options:
+            hass.config_entries.async_update_entry(
+                entry,
+                data={
+                    CONF_DISCOVERY: entry.data.get(CONF_DISCOVERY),
+                    CONF_NAME: entry.data.get(CONF_NAME),
+                    CONF_IP_ADDRESS: entry.data[CONF_IP_ADDRESS],
+                },
+                options={
+                    CONF_MODEL: entry.data.get(CONF_MODEL, ""),
+                    CONF_TRANSITION: entry.data.get(
+                        CONF_TRANSITION, DEFAULT_TRANSITION
+                    ),
+                    CONF_MODE_MUSIC: entry.data.get(
+                        CONF_MODE_MUSIC, DEFAULT_MODE_MUSIC
+                    ),
+                    CONF_SAVE_ON_CHANGE: entry.data.get(
+                        CONF_SAVE_ON_CHANGE, DEFAULT_SAVE_ON_CHANGE
+                    ),
+                    CONF_NIGHTLIGHT_SWITCH: entry.data.get(
+                        CONF_NIGHTLIGHT_SWITCH, DEFAULT_NIGHTLIGHT_SWITCH
+                    ),
+                },
+            )
 
-    if DOMAIN in config:
-        for ipaddr, device_config in conf[CONF_DEVICES].items():
-            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
-            _setup_device(hass, config, ipaddr, device_config)
+        config = {
+            CONF_NAME: entry.data.get(CONF_NAME),  # Support name import from yaml
+            **entry.options,
+        }
+        await _async_setup_device(hass, entry.data[CONF_IP_ADDRESS], config)
+        unset = entry.add_update_listener(_async_update_listener)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = {
+            DATA_UNSUB_UPDATE_LISTENER: unset,
+        }
+        for component in PLATFORMS:
```

Comment:
Many lines are duplicated from above in the discovery case. Can we consolidate it to make it more DRY? Why not use `_initialize` in both cases? Have it accept any variation as parameters.